src/server/src/main/java/io/cassandrareaper/jmx/JmxConnectionFactory.java

```diff
@@ -163,4 +175,55 @@ private void decrementSuccessfullConnections(String host) {
       LOG.warn("Could not decrement JMX successfull connections counter for host {}", host, e);
     }
   }
+
+
+
+  public void initializeJmxSeeds(Cluster cluster) {
+    LOG.info("Initializing JMX seed list for cluster {}...", cluster.getName());
+    List<Callable<Optional<String>>> jmxTasks = Lists.newArrayList();
+    List<String> seedHosts = Lists.newArrayList();
+    seedHosts.addAll(cluster.getSeedHosts());
+
+    for (int i = 0; i < seedHosts.size(); i++) {
+      jmxTasks.add(connectToJmx(Arrays.asList(seedHosts.get(i))));
+      if (i % 10 == 0 || i == seedHosts.size() - 1) {
+        tryConnectingToJmxSeeds(jmxTasks);
+        jmxTasks = Lists.newArrayList();
+      }
+    }
+  }
+
+  private Callable<Optional<String>> connectToJmx(List<String> endpoints) {
+    return () -> {
+      try (JmxProxy jmxProxy =
+          this.connectAny(Optional.absent(), endpoints, JmxProxy.JMX_CONNECTION_TIMEOUT)) {
+
+        return Optional.of(endpoints.get(0));
+
+      } catch (RuntimeException e) {
+        LOG.debug("failed to connect to hosts {} through JMX", endpoints.get(0), e);
+        return Optional.absent();
+      }
+    };
+  }
```

Comment:
I don't understand this method…

why couldn't it be:
```
  private Callable<Optional<String>> testJmxConnection(String endpoint) {
    return () -> {
      try (JmxProxy jmxProxy = connect(Optional.absent(), endpoint, JmxProxy.JMX_CONNECTION_TIMEOUT)) {
        incrementSuccessfullConnections(endpoint);
        return Optional.of(endpoint);
      } catch (ReaperException | RuntimeException e) {
        decrementSuccessfullConnections(host);
        LOG.debug("failed to connect to hosts {} through JMX", endpoints.get(0), e);
        return Optional.absent();
      }
    };
  }
```

Reply:
I tried to avoid duplicating the logic of maintaining the counters for successful connections which is already implemented in `connectAny()`.

Reply:
Could we encapsulate too these counters into a separate class, like…
```
static final class HostConnectionsInfo {

  private final ConcurrentMap<String, AtomicInteger> connectionCounts = Maps.newConcurrentMap();

  void incrementSuccessfullConnections(String host) {
    try {
      AtomicInteger successes = connectionCounts.putIfAbsent(host, new AtomicInteger(1));
      if (null != successes && successes.get() <= 20) {
        successes.incrementAndGet();
      }
      LOG.debug("Host {} has {} successfull connections", host, successes);
      if (null != metricRegistry) {
        metricRegistry
            .counter(MetricRegistry.name(JmxConnectionFactory.class, "connections", host.replace('.', '-')))
            .inc();
      }
    } catch (RuntimeException e) {
      LOG.warn("Could not increment JMX successfull connections counter for host {}", host, e);
    }
  }

  void decrementSuccessfullConnections(String host) {
    try {
      AtomicInteger successes = connectionCounts.putIfAbsent(host, new AtomicInteger(-1));
      if (null != successes && successes.get() >= -5) {
        successes.decrementAndGet();
      }
      LOG.debug("Host {} has {} successfull connections", host, successes);
      if (null != metricRegistry) {
        metricRegistry
            .counter(MetricRegistry.name(JmxConnectionFactory.class, "connections", host.replace('.', '-')))
            .dec();
      }
    } catch (RuntimeException e) {
      LOG.warn("Could not decrement JMX successfull connections counter for host {}", host, e);
    }
  }
}
```

Re-using the logic: which is only calling the increment/decrement methods, makes more sense then.

Reply:
I've moved the counters into a separate class as suggested.

Looks good ?